File: Task_1/c/stats_utils.c
/*
 * Statistical Utilities for C
 */
#include <math.h>
#include "stats_utils.h"

/**
 * Calculate average of an array
 */
double calculate_average(double* values, int length) {
    double sum = 0;
    for (int i = 0; i < length; i++) {
        sum += values[i];
    }
    return sum / length;
}

/**
 * Find minimum value in array
 */
double find_min(double* values, int length) {
    double min = values[0];
    for (int i = 1; i < length; i++) {
        if (values[i] < min) min = values[i];
    }
    return min;
}

/**
 * Find maximum value in array
 */
double find_max(double* values, int length) {
    double max = values[0];
    for (int i = 1; i < length; i++) {
        if (values[i] > max) max = values[i];
    }
    return max;
}

/**
 * Calculate standard deviation
 */
double calculate_std_dev(double* values, int length) {
    if (length <= 1) return 0;
    
    double mean = calculate_average(values, length);
    double sum_squared_diffs = 0;
    
    for (int i = 0; i < length; i++) {
        double diff = values[i] - mean;
        sum_squared_diffs += diff * diff;
    }
    
    return sqrt(sum_squared_diffs / (length - 1));
}

File: Task_1/code/java/StatsUtils.java
/**
 * Statistical Utilities for Java
 */
public class StatsUtils {
    
    /**
     * Calculate average of an array
     */
    public static double calculateAverage(double[] values) {
        double sum = 0;
        for (double value : values) {
            sum += value;
        }
        return sum / values.length;
    }
    
    /**
     * Find minimum value in array
     */
    public static double findMin(double[] values) {
        double min = values[0];
        for (double value : values) {
            if (value < min) min = value;
        }
        return min;
    }
    
    /**
     * Find maximum value in array
     */
    public static double findMax(double[] values) {
        double max = values[0];
        for (double value : values) {
            if (value > max) max = value;
        }
        return max;
    }
    
    /**
     * Calculate standard deviation
     */
    public static double calculateStdDev(double[] values) {
        if (values.length <= 1) return 0;
        
        double mean = calculateAverage(values);
        double sumSquaredDiffs = 0;
        
        for (double value : values) {
            double diff = value - mean;
            sumSquaredDiffs += diff * diff;
        }
        
        return Math.sqrt(sumSquaredDiffs / (values.length - 1));
    }
}